File: Engine/Android/core/src/test/kotlin/despia/engine/MotionConformanceTest.kt
package despia.engine

import java.io.File
import kotlin.math.abs
import kotlin.math.sqrt
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import org.junit.jupiter.api.DynamicTest
import org.junit.jupiter.api.TestFactory

/**
 * The SHARED UI MOTION corpus (OpenSource/Conformance/motion/
 * {curves,spring,retarget,physics}.json) through THIS runtime's motion kernel — the
 * Kotlin leg of ui-motion.md. The TS reference (packages/kernel/test/
 * motion-conformance.test.ts) and the Swift twin (record lane, MotionConformance) run the
 * SAME files, so `anim="spring"` can never again mean three different curves on the three
 * renderers. Expected numbers were computed by an independent scratch derivation, never
 * by any kernel under test.
 *
 * The Compose edge (:render StackMotion.kt) derives its AnimationSpec numbers from the
 * MotionSpec this test pins, so a green run here is a green run for the actual UI.
 */
class MotionConformanceTest {

    private val tolerance = 1.5e-6

    private fun corpusDir(): File {
        var dir = File(System.getProperty("user.dir") ?: ".").absoluteFile
        while (true) {
            val candidate = File(dir, "OpenSource/Conformance/motion")
            if (candidate.isDirectory) return candidate
            dir = dir.parentFile ?: error("OpenSource/Conformance/motion not found")
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun load(file: String): Map<String, Any?> =
        json(File(corpusDir(), file).readText()).foundationValue as? Map<String, Any?>
            ?: error("$file: not a JSON object")

    @Suppress("UNCHECKED_CAST")
    private fun rows(doc: Map<String, Any?>, key: String, minimum: Int): List<Map<String, Any?>> {
        val list = doc[key] as? List<Map<String, Any?>> ?: error("no $key[]")
        assertTrue(list.size >= minimum, "motion/$key is suspiciously small (${list.size})")
        return list
    }

    private fun num(v: Any?): Double = (v as Number).toDouble()
    private fun str(v: Any?): String? = v as? String

    private fun close(actual: Double, expected: Double, label: String) {
        assertTrue(abs(actual - expected) <= tolerance, "$label: $actual !~ $expected")
    }

    // ── curves.json ──────────────────────────────────────────────────────────────────

    @TestFactory
    @Suppress("UNCHECKED_CAST")
    fun curvesParse(): List<DynamicTest> {
        val doc = load("curves.json")
        return rows(doc, "parse", 15).map { c ->
            val name = str(c["name"]) ?: "?"
            DynamicTest.dynamicTest("motion/curves parse — $name") {
                val codes = mutableListOf<String>()
                val spec = Motion.parse(str(c["anim"]), str(c["animDuration"])) { code, _ -> codes.add(code) }
                val expect = c["expect"] as Map<String, Any?>
                assertEquals(str(expect["name"]), spec.name, "name")
                assertEquals(str(expect["kind"]), if (spec.isSpring) "spring" else "curve", "kind")
                close(spec.durationMs, num(expect["durationMs"]), "durationMs")
                close(Motion.settleMs(spec), num(expect["durationMs"]), "settleMs")
                if (spec.isSpring) {
                    close(spec.response, num(expect["response"]), "response")
                    close(spec.dampingFraction, num(expect["dampingFraction"]), "dampingFraction")
                } else {
                    close(spec.x1, num(expect["x1"]), "x1")
                    close(spec.y1, num(expect["y1"]), "y1")
                    close(spec.x2, num(expect["x2"]), "x2")
                    close(spec.y2, num(expect["y2"]), "y2")
                }
                val wanted = (c["diagnostics"] as List<Any?>).joinToString(",") { str(it) ?: "" }
                assertEquals(wanted, codes.joinToString(","), "diagnostics")
            }
        }
    }

    @TestFactory
    @Suppress("UNCHECKED_CAST")
    fun curvesPresets(): List<DynamicTest> {
        val presets = load("curves.json")["presets"] as Map<String, Map<String, Any?>>
        val table = mapOf(
            "keep" to Motion.PRESET_KEEP,
            "press" to Motion.PRESET_PRESS,
            "default" to Motion.PRESET_DEFAULT,
        )
        return presets.entries.map { (key, expect) ->
            DynamicTest.dynamicTest("motion/curves preset — $key") {
                val spec = table[key] ?: error("no preset for $key")
                assertEquals(str(expect["name"]), spec.name, "$key: name")
                assertEquals(str(expect["kind"]), if (spec.isSpring) "spring" else "curve", "$key: kind")
                close(spec.durationMs, num(expect["durationMs"]), "$key: durationMs")
                close(spec.x1, num(expect["x1"]), "$key: x1")
                close(spec.y1, num(expect["y1"]), "$key: y1")
                close(spec.x2, num(expect["x2"]), "$key: x2")
                close(spec.y2, num(expect["y2"]), "$key: y2")
            }
        }
    }

    @TestFactory
    @Suppress("UNCHECKED_CAST")
    fun curvesProgress(): List<DynamicTest> {
        val doc = load("curves.json")
        return rows(doc, "progress", 8).map { c ->
            val name = str(c["name"]) ?: "?"
            DynamicTest.dynamicTest("motion/curves progress — $name") {
                val spec = Motion.parse(str(c["anim"]), str(c["animDuration"]))
                close(spec.durationMs, num(c["durationMs"]), "durationMs")
                for (s in c["samples"] as List<Map<String, Any?>>) {
                    close(Motion.progress(spec, num(s["t"])), num(s["p"]), "p(${num(s["t"])})")
                }
            }
        }
    }

    // ── spring.json — the crux of the 1:1 claim ──────────────────────────────────────

    @TestFactory
    @Suppress("UNCHECKED_CAST")
    fun springConversion(): List<DynamicTest> {
        val doc = load("spring.json")
        return rows(doc, "conversion", 5).map { c ->
            val name = str(c["name"]) ?: "?"
            DynamicTest.dynamicTest("motion/spring conversion — $name") {
                val response = num(c["response"])
                val dampingFraction = num(c["dampingFraction"])
                val k = Motion.springConstants(response, dampingFraction)
                close(sqrt(k[0]), num(c["omega"]), "omega")
                close(k[0], num(c["stiffness"]), "stiffness")
                close(k[1], num(c["damping"]), "damping")
                close(k[1] / (2.0 * sqrt(k[0])), num(c["zeta"]), "zeta")
                // At the authored damping fraction (0.8 — the only one `anim=` reaches),
                // the parse must produce exactly these numbers and this settle time.
                if (dampingFraction == 0.8) {
                    val spec = Motion.parse("spring", response.toString())
                    assertTrue(spec.isSpring, "spring")
                    close(spec.stiffness, num(c["stiffness"]), "spec.stiffness")
                    close(spec.damping, num(c["damping"]), "spec.damping")
                    close(spec.durationMs, num(c["settleMs"]), "settleMs")
                }
            }
        }
    }

    @TestFactory
    @Suppress("UNCHECKED_CAST")
    fun springProgress(): List<DynamicTest> {
        val doc = load("spring.json")
        return rows(doc, "progress", 4).map { c ->
            val name = str(c["name"]) ?: "?"
            DynamicTest.dynamicTest("motion/spring progress — $name") {
                // Non-default damping fractions are not reachable through `anim=`; build
                // from the authoring plane so the corpus pins the whole family.
                val spec = Motion.spring(num(c["response"]), num(c["dampingFraction"]))
                close(Motion.settleMs(spec), num(c["settleMs"]), "settleMs")
                for (s in c["samples"] as List<Map<String, Any?>>) {
                    close(Motion.progress(spec, num(s["t"])), num(s["p"]), "p(${num(s["t"])})")
                }
            }
        }
    }

    // ── retarget.json — the interruption law ─────────────────────────────────────────

    @TestFactory
    @Suppress("UNCHECKED_CAST")
    fun retarget(): List<DynamicTest> {
        val doc = load("retarget.json")
        return rows(doc, "cases", 6).map { c ->
            val name = str(c["name"]) ?: "?"
            DynamicTest.dynamicTest("motion/retarget — $name") {
                val spec = Motion.parse(str(c["anim"]), str(c["animDuration"]))
                var state = Motion.start(num(c["from"]), num(c["to"]), 0.0)
                val events = (c["events"] as List<Map<String, Any?>>).sortedBy { num(it["at"]) }
                var next = 0
                for (s in c["samples"] as List<Map<String, Any?>>) {
                    val t = num(s["t"])
                    while (next < events.size && num(events[next]["at"]) <= t) {
                        state = Motion.retarget(spec, state, num(events[next]["to"]), num(events[next]["at"]))
                        next += 1
                    }
                    val got = Motion.value(spec, state, t)
                    close(got.value, num(s["value"]), "value($t)")
                    assertEquals(s["done"] as Boolean, got.done, "done($t)")
                }
            }
        }
    }

    // ── physics.json — the UI physics primitives ─────────────────────────────────────

    @TestFactory
    @Suppress("UNCHECKED_CAST")
    fun physics(): List<DynamicTest> {
        val doc = load("physics.json")
        val out = mutableListOf<DynamicTest>()

        val constants = doc["constants"] as Map<String, Any?>
        out.add(DynamicTest.dynamicTest("motion/physics — the pinned constants") {
            close(Motion.DECELERATION_RATE, num(constants["decelerationRate"]), "decelerationRate")
            close(Motion.DECAY_TAU_MS, num(constants["tauMs"]), "tauMs")
            close(Motion.DECAY_MIN_VELOCITY, num(constants["minVelocity"]), "minVelocity")
            close(Motion.RUBBER_BAND_C, num(constants["rubberBandC"]), "rubberBandC")
            assertTrue(Motion.RUBBER_BAND_RELEASE.isSpring, "the release spec is a spring")
            close(Motion.RUBBER_BAND_RELEASE.response, num(constants["releaseResponse"]), "releaseResponse")
            close(
                Motion.RUBBER_BAND_RELEASE.dampingFraction,
                num(constants["releaseDampingFraction"]), "releaseDampingFraction",
            )
        })

        for (c in rows(doc, "decay", 3)) {
            out.add(DynamicTest.dynamicTest("motion/physics decay — ${str(c["name"])}") {
                val x0 = num(c["x0"])
                val v0 = num(c["v0"])
                close(Motion.decayTarget(x0, v0), num(c["target"]), "target")
                close(Motion.decayDurationMs(v0), num(c["durationMs"]), "durationMs")
                for (s in c["samples"] as List<Map<String, Any?>>) {
                    close(Motion.decayAt(x0, v0, num(s["t"])), num(s["x"]), "x(${num(s["t"])})")
                }
            })
        }

        for (c in rows(doc, "rubberBand", 6)) {
            out.add(DynamicTest.dynamicTest("motion/physics rubberBand — ${str(c["name"])}") {
                val d = num(c["dimension"])
                val y = Motion.rubberBand(num(c["x"]), d)
                close(y, num(c["y"]), "y")
                close(Motion.rubberBandInverse(y, d), num(c["inverse"]), "inverse round trip")
            })
        }

        for (c in rows(doc, "snap", 6)) {
            out.add(DynamicTest.dynamicTest("motion/physics snap — ${str(c["name"])}") {
                val points = (c["points"] as List<Any?>).map { num(it) }
                val got = Motion.snapTarget(num(c["x0"]), num(c["v0"]), points)
                close(got.projected, num(c["projected"]), "projected")
                close(got.target, num(c["target"]), "target")
                assertEquals((num(c["index"])).toInt(), got.index, "index")
            })
        }

        return out
    }
}

File: Web/packages/kernel/test/motion-conformance.test.ts
//
//  motion-conformance.test.ts — the SHARED UI MOTION corpus
//  (OpenSource/Conformance/motion/{curves,spring,retarget,physics}.json) through the TS
//  motion kernel — the REFERENCE leg of ui-motion.md. The Kotlin twin
//  (:core MotionConformanceTest) and the Swift twin (record lane, MotionConformance) run
//  the SAME files, so `anim="spring"` cannot mean three different curves again. Expected
//  numbers were computed by an independent scratch derivation, never by this kernel.
//

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync, existsSync } from "node:fs";
import { join, dirname, resolve } from "node:path";

import {
  parseMotion, motionProgress, motionSettleMs, motionSpringConstants,
  motionStart, motionValue, motionRetarget, motionSpring,
  decayAt, decayTarget, decayDurationMs, rubberBand, rubberBandInverse, snapTarget,
  MOTION_PRESET_KEEP, MOTION_PRESET_PRESS, MOTION_PRESET_DEFAULT,
  MOTION_DECELERATION_RATE, MOTION_DECAY_TAU_MS, MOTION_DECAY_MIN_VELOCITY,
  MOTION_RUBBER_BAND_C, MOTION_RUBBER_BAND_RELEASE,
  type MotionSpec,
} from "../src/index.ts";

const TOLERANCE = 1.5e-6;

function corpusDir(): string {
  let dir = resolve(import.meta.dirname ?? ".");
  for (;;) {
    const candidate = join(dir, "OpenSource/Conformance/motion");
    if (existsSync(candidate)) return candidate;
    const parent = dirname(dir);
    if (parent === dir) throw new Error("motion corpus not found");
    dir = parent;
  }
}

function load<T>(file: string): T {
  return JSON.parse(readFileSync(join(corpusDir(), file), "utf-8")) as T;
}

function close(actual: number, expected: number, label: string): void {
  assert.ok(Math.abs(actual - expected) <= TOLERANCE, `${label}: ${actual} !~ ${expected}`);
}

// ── curves.json ──────────────────────────────────────────────────────────────────────

type ParseCase = {
  name: string; anim: string | null; animDuration: string | null;
  expect: {
    name: string; kind: string; durationMs: number;
    x1?: number; y1?: number; x2?: number; y2?: number;
    response?: number; dampingFraction?: number;
  };
  diagnostics: string[];
};
type ProgressCase = {
  name: string; anim: string | null; animDuration: string | null; durationMs: number;
  samples: { t: number; p: number }[];
};
type PresetShape = { name: string; kind: string; durationMs: number; x1: number; y1: number; x2: number; y2: number };
type CurvesDoc = { presets: Record<string, PresetShape>; parse: ParseCase[]; progress: ProgressCase[] };

const curves = load<CurvesDoc>("curves.json");
assert.ok(curves.parse.length >= 15, "motion/curves parse corpus is suspiciously small");
assert.ok(curves.progress.length >= 8, "motion/curves progress corpus is suspiciously small");

for (const c of curves.parse) {
  test(`motion/curves parse — ${c.name}`, () => {
    const codes: string[] = [];
    const spec = parseMotion(c.anim, c.animDuration, (d) => codes.push(d.code));
    assert.equal(spec.kind, c.expect.kind, "kind");
    assert.equal(spec.name, c.expect.name, "name");
    close(spec.durationMs, c.expect.durationMs, "durationMs");
    close(motionSettleMs(spec), c.expect.durationMs, "motionSettleMs");
    if (spec.kind === "spring") {
      close(spec.response, c.expect.response!, "response");
      close(spec.dampingFraction, c.expect.dampingFraction!, "dampingFraction");
    } else {
      assert.deepEqual([spec.x1, spec.y1, spec.x2, spec.y2],
        [c.expect.x1, c.expect.y1, c.expect.x2, c.expect.y2], "control points");
    }
    assert.deepEqual(codes, c.diagnostics, "diagnostics");
  });
}

test("motion/curves — the pinned presets", () => {
  const table: Record<string, MotionSpec> = {
    keep: MOTION_PRESET_KEEP, press: MOTION_PRESET_PRESS, default: MOTION_PRESET_DEFAULT,
  };
  for (const [key, expect] of Object.entries(curves.presets)) {
    const spec = table[key];
    assert.ok(spec !== undefined, `no preset for ${key}`);
    assert.equal(spec.kind, expect.kind, `${key}: kind`);
    assert.equal(spec.name, expect.name, `${key}: name`);
    close(spec.durationMs, expect.durationMs, `${key}: durationMs`);
    assert.equal(spec.kind, "curve");
    if (spec.kind === "curve") {
      assert.deepEqual([spec.x1, spec.y1, spec.x2, spec.y2],
        [expect.x1, expect.y1, expect.x2, expect.y2], `${key}: control points`);
    }
  }
});

for (const c of curves.progress) {
  test(`motion/curves progress — ${c.name}`, () => {
    const spec = parseMotion(c.anim, c.animDuration);
    close(spec.durationMs, c.durationMs, "durationMs");
    for (const s of c.samples) close(motionProgress(spec, s.t), s.p, `p(${s.t})`);
  });
}

// ── spring.json ──────────────────────────────────────────────────────────────────────

type ConversionCase = {
  name: string; response: number; dampingFraction: number;
  omega: number; stiffness: number; damping: number; zeta: number; settleMs: number;
};
type SpringProgressCase = {
  name: string; response: number; dampingFraction: number; settleMs: number;
  samples: { t: number; p: number }[];
};
const spring = load<{ conversion: ConversionCase[]; progress: SpringProgressCase[] }>("spring.json");
assert.ok(spring.conversion.length >= 5, "motion/spring conversion corpus is suspiciously small");
assert.ok(spring.progress.length >= 4, "motion/spring progress corpus is suspiciously small");

for (const c of spring.conversion) {
  test(`motion/spring conversion — ${c.name}`, () => {
    const { stiffness, damping } = motionSpringConstants(c.response, c.dampingFraction);
    close(Math.sqrt(stiffness), c.omega, "omega");
    close(stiffness, c.stiffness, "stiffness");
    close(damping, c.damping, "damping");
    close(damping / (2 * Math.sqrt(stiffness)), c.zeta, "zeta");
    // At the authored damping fraction (0.8 — the only one `anim=` can reach), the
    // parse must produce exactly these numbers and this settle time.
    if (c.dampingFraction === 0.8) {
      const spec = parseMotion("spring", String(c.response));
      assert.equal(spec.kind, "spring");
      if (spec.kind === "spring") {
        close(spec.stiffness, c.stiffness, "spec.stiffness");
        close(spec.damping, c.damping, "spec.damping");
      }
      close(spec.durationMs, c.settleMs, "settleMs");
    }
  });
}

for (const c of spring.progress) {
  test(`motion/spring progress — ${c.name}`, () => {
    // Non-default damping fractions are not reachable through `anim=`; build the spec
    // from the authoring plane so the corpus can pin the whole oscillator family.
    const pinned = motionSpring(c.response, c.dampingFraction);
    close(motionSettleMs(pinned), c.settleMs, "settleMs");
    for (const s of c.samples) close(motionProgress(pinned, s.t), s.p, `p(${s.t})`);
  });
}

// ── retarget.json ────────────────────────────────────────────────────────────────────

type RetargetCase = {
  name: string; anim: string | null; animDuration: string | null;
  from: number; to: number;
  events: { at: number; to: number }[];
  samples: { t: number; value: number; done: boolean }[];
};
const retarget = load<{ cases: RetargetCase[] }>("retarget.json");
assert.ok(retarget.cases.length >= 6, "motion/retarget corpus is suspiciously small");

for (const c of retarget.cases) {
  test(`motion/retarget — ${c.name}`, () => {
    const spec = parseMotion(c.anim, c.animDuration);
    let state = motionStart(c.from, c.to, 0);
    const pending = [...c.events].sort((a, b) => a.at - b.at);
    let next = 0;
    for (const s of c.samples) {
      while (next < pending.length && pending[next]!.at <= s.t) {
        state = motionRetarget(spec, state, pending[next]!.to, pending[next]!.at);
        next += 1;
      }
      const got = motionValue(spec, state, s.t);
      close(got.value, s.value, `value(${s.t})`);
      assert.equal(got.done, s.done, `done(${s.t})`);
    }
  });
}

// ── physics.json ─────────────────────────────────────────────────────────────────────

type PhysicsDoc = {
  constants: {
    decelerationRate: number; tauMs: number; minVelocity: number; rubberBandC: number;
    releaseResponse: number; releaseDampingFraction: number;
  };
  decay: { name: string; x0: number; v0: number; target: number; durationMs: number; samples: { t: number; x: number }[] }[];
  rubberBand: { name: string; x: number; dimension: number; y: number; inverse: number }[];
  snap: { name: string; x0: number; v0: number; points: number[]; projected: number; target: number; index: number }[];
};
const physics = load<PhysicsDoc>("physics.json");
assert.ok(physics.decay.length >= 3 && physics.rubberBand.length >= 6 && physics.snap.length >= 6,
  "motion/physics corpus is suspiciously small");

test("motion/physics — the pinned constants", () => {
  close(MOTION_DECELERATION_RATE, physics.constants.decelerationRate, "decelerationRate");
  close(MOTION_DECAY_TAU_MS, physics.constants.tauMs, "tauMs");
  close(MOTION_DECAY_MIN_VELOCITY, physics.constants.minVelocity, "minVelocity");
  close(MOTION_RUBBER_BAND_C, physics.constants.rubberBandC, "rubberBandC");
  assert.equal(MOTION_RUBBER_BAND_RELEASE.kind, "spring");
  if (MOTION_RUBBER_BAND_RELEASE.kind === "spring") {
    close(MOTION_RUBBER_BAND_RELEASE.response, physics.constants.releaseResponse, "releaseResponse");
    close(MOTION_RUBBER_BAND_RELEASE.dampingFraction, physics.constants.releaseDampingFraction,
      "releaseDampingFraction");
  }
});

for (const c of physics.decay) {
  test(`motion/physics decay — ${c.name}`, () => {
    close(decayTarget(c.x0, c.v0), c.target, "target");
    close(decayDurationMs(c.v0), c.durationMs, "durationMs");
    for (const s of c.samples) close(decayAt(c.x0, c.v0, s.t), s.x, `x(${s.t})`);
  });
}

for (const c of physics.rubberBand) {
  test(`motion/physics rubberBand — ${c.name}`, () => {
    const y = rubberBand(c.x, c.dimension);
    close(y, c.y, "y");
    close(rubberBandInverse(y, c.dimension), c.inverse, "inverse round trip");
  });
}

for (const c of physics.snap) {
  test(`motion/physics snap — ${c.name}`, () => {
    const got = snapTarget(c.x0, c.v0, c.points);
    close(got.projected, c.projected, "projected");
    close(got.target, c.target, "target");
    assert.equal(got.index, c.index, "index");
  });
}
